### PR TITLE
Improve high frequency handling on the graph

### PR DESCRIPTION
**Pull request by jamorham**
**Problem this PR is to solve:**

In order to be able to support higher frequency data sources (eg 1 minute per reading) there are internals within xDrip which need to change.

**What this PR does:**

In order to avoid problems with graph resolution and fuzzing of the value timestamps, a new class is introduced which is better able to cope with higher resolution timestamps and a reduced fuzzing step of 30 seconds. Compatibility with existing subsystems should be maintained.

Some internal variables are converted from double to long to better represent timestamps and some processed variables representing coordinates are expressed in doubles and as offsets from an origin point.

**What this PR doesn't do:**

This doesn't enable facilities for higher frequency data sources, it just adjusts some incompatible components so that they will be able to support higher frequencies when other changes are made to facilitate that.

Additionally I've reworked some code relating to viewport handling to simplify it and also to solve the issue where the selected portion of the graph would jump to the most recent reading when it arrived while a user may be in the middle of panning back to an earlier time to review data. Now this wont happen if the viewport has been manually moved within the last 45 seconds.

**What needs to be tested:**

I've provided this PR so that interested parties can test out these changes and confirm that nothing relating to the existing behavior is broken by this. This is quite a significant change and because of how xDrip optimizes calculations by combining graph rendering and analysis there is plenty of scope for things to go wrong.

Internally I have done some testing with high frequency data sources (1 minute reading) and this does seem to resolve the issue and work well, but those tests have been limited and it would be better if more developers try this with existing 5 minute data sources to ensure the change is safe and reliable.

**Time scale:**

I would like to be able to merge this around the 19th of Aug 2022

### CODE REVIEW

**Review by tzachi-dar — COMMENTED** [review]
Generally seems OK with me.
Please see one more comment.

**Comment by Navid200** [comment]
I have been using this for the past 8 days.
I haven't seen anything wrong expect smart alert misbehaving.  But, I don't think that has anything to do with this change.  
It is possible that it is a coincidence that I have noticed it now.  We already have an open issue for smart alert:  https://github.com/NightscoutFoundation/xDrip/issues/1786

File: app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/HPointValue.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * JamOrHam
+ * <p>
+ * Due to range issues with float we need to wrap around a friendly origin timestamp expressing
+ * but storing real value in a bigger primitive type while maintaining compatibility with libraries
+ */
+
+@AllArgsConstructor
+public class HPointValue extends PointValue {
+
+    private static final long OFFSET = (BuildConfig.buildTimestamp - Constants.MONTH_IN_MS) / FUZZER;
```
[the remark this answers is not in the document]
Comment (jamorham) [reply app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/HPointValue.java:20]: Yes values will be negative but that will be unlikely to present a problem as the offset will be reversed whenever the value is used. This should be tested of course...

**Comment by jamorham** [comment]
@Navid200 I don't think smart alert can be affected by this as this should only be affecting graphs.

I plan to merge this shortly.

**Comment by 4RK4N** [comment]
seems to have broken the broadcastserviceAPI. bg data is recieved but the graph is not. it worked before the nightly with this PR included

**Pull request merged**